tests: Rework tests and enable integration tests (#2120)
elastisys/compliantkubernetes-apps: tests/README.md

# Apps Tests

The test suite is implemented using [`bats`](https://github.com/bats-core/bats-core) and [`cypress`](https://github.com/cypress-io/cypress), with unit, regression, integration, and end-to-end tests under their own respective directory.

> Generators are employed to generate `bats` tests from `cypress` and `gotmpl` tests to integrate it into the rest of test suites.

Reviewer: This was a bit hard to parse. What about something like this?
```suggestion
For each `cypress` test a `bats` test is generated that runs the `cypress` test and verifies the result.
Some tests are also generated from `gotmpl` templates.
```

Reviewer: On the other hand there is already good descriptions of the different tests that you can write and the code generation further down so maybe this can be removed?

The test harness is implemented to be run in a container using either rootful `docker` or rootless `podman`.

## Usage

> [!note]
> All instructions assume that you are standing in the `tests/` directory.

> [!warning]
> Do not use `docker` or `podman` directly from tests, as they take differing arguments and may depend on variables present on the host not available within the tests container.
> Instead use tools like `buildah` for building and pushing or `skopeo` for syncing and pulling.
>
> Additionally tests running in the test container might hang on interrupts, requiring the container to be killed.

The tests differentiate between static and dynamic tests, all static tests can be run without setting up an environment, and all dynamic tests requires an environment to test.
Static tests are tagged with `static`.

> [!danger]

> > This type of distinction will be phased out and all unit, regression, and integration tests will be required to be static!

Reviewer: This doesn't seem to render when I preview the rendered markdown

```bash
# build and run
make build-unit
make run-unit
```

The `tests / integration` workflow on GitHub is invoked with the following commands:

```bash
# prepare local cache and local resolve:
../scripts/local-cluster.sh cache create
../scripts/local-cluster.sh resolve create integration.dev-ck8s.com
# build and run
make build-main
make run-integration
```

### Usage with Makefile

> [!note]
> The container might struggle to prompt for kube-login and gpg-agent, but if those are activated before by accessing the clusters and using gpg then the session can be reused.
> This is applicable for regression, integration, and end-to-end tests.

You must have `make` and either rootful `docker` or rootless `podman` installed!
Check the [DEVELOPMENT](../DEVELOPMENT.md) docs additional requirements to run local-clusters for integration and regression tests.

Run all tests:

```bash
make run-all
```

Run selected tests:

```bash
make run-<target>                # list targets with make list
make run-<target>/<suite>        # list targets with make list-<target>
make run-<target>/<suite>/<file> # list targets with make list-<target>/<suite>
```

> To force cypress and template tests to be regenerated run:

Author: Reworded with an if...

> To force cypress and template tests to be regenerated run:

Reviewer: It reads to me that tests are automatically generated, is that really the case? For example, if I add a new test do I not have to also generate the bats wrapper for it myself?

> To force cypress and template tests to be regenerated run:

Author: Yes, if you add a cypress or template then the bats files are generated from them automatically if a source file is newer than the destination file.

> To force cypress and template tests to be regenerated run:

Reviewer: When? Clearly not as soon as I save the file? I assume I have to do _something_ to generate the new test files before I commit my changes to the git repository?

> To force cypress and template tests to be regenerated run:

Author: As a regular make prerequisite.

> To force cypress and template tests to be regenerated run:

Reviewer: Ah, that's prefect!

Additionally tests can be filtered via tags using the `-<tags,...>` suffix to the `run-<target>` command.

Clean up:

```bash
# remove dependencies and generated files
make clean

# remove generated files
make clean-gen

> ```

Reviewer: Can this entire section not be removed now when we commit the generated tests? Or is there still a good reason to clean things up sometimes?

> Additionally tests can be filtered via tags using the `--filter-tags <tags,...>` argument.

Reviewer: This looks odd when it's above the `cypress` instructions. Maybe add a sentence similar to the one above the `bats` instructions or move this?
